opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java

```diff
@@ -62,28 +84,39 @@ public OnmsAlarm persist(Event event) {
         if (!checkEventSanityAndDoWeProcess(event)) {
             return null;
         }
-        LOG.debug("process: {}; nodeid: {}; ipaddr: {}; serviceid: {}", event.getUei(), event.getNodeid(), event.getInterface(), event.getService());
 
-        return addOrReduceEventAsAlarm(event);
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("process: {}; nodeid: {}; ipaddr: {}; serviceid: {}", event.getUei(), event.getNodeid(), event.getInterface(), event.getService());
+        }
+
+        // Process the alarm inside a transaction
+        OnmsAlarmAndLifecycleEvent alarmAndEvent = m_transactionOperations.execute(new TransactionCallback<OnmsAlarmAndLifecycleEvent>() {
+            @Override
+            public OnmsAlarmAndLifecycleEvent doInTransaction(TransactionStatus arg0) {
+                return addOrReduceEventAsAlarm(event);
+            }
+        });
+
+        // Send the event outside of the database transaction
+        m_eventForwarder.sendNow(alarmAndEvent.getEvent());
+
+        return alarmAndEvent.getAlarm();
     }
 
-    private OnmsAlarm addOrReduceEventAsAlarm(Event event) {
-        //TODO: Understand why we use Assert
-        Assert.notNull(event, "Incoming event was null, aborting"); 
-        Assert.isTrue(event.getDbid() > 0, "Incoming event has an illegal dbid (" + event.getDbid() + "), aborting");
-        
-        
-        //for some reason when we get here the event from the DB doesn't have the LogMsg (in my tests anyway)
+    private OnmsAlarmAndLifecycleEvent addOrReduceEventAsAlarm(Event event) {
+        // 2012-03-11 pbrane: for some reason when we get here the event from the DB doesn't have the LogMsg (in my tests anyway)
         OnmsEvent e = m_eventDao.get(event.getDbid());
         Assert.notNull(e, "Event was deleted before we could retrieve it and create an alarm.");
-    
+
         String reductionKey = event.getAlarmData().getReductionKey();
         LOG.debug("addOrReduceEventAsAlarm: looking for existing reduction key: {}", reductionKey);
         OnmsAlarm alarm = m_alarmDao.findByReductionKey(reductionKey);
 
         EventBuilder ebldr = null;
         if (alarm == null) {
-            LOG.debug("addOrReduceEventAsAlarm: reductionKey:{} not found, instantiating new alarm", reductionKey);
+            if (LOG.isDebugEnabled()) {
```

Comment:
I was changing all of the debug() calls to avoid the overhead of creating the arguments but in this particular log message, there is no extra overhead since the value is an existing variable. I guess I was overzealous. :)